pkg/logger/filter.go

```diff
@@ -0,0 +1,337 @@
+package logger
+
+import (
+	"regexp"
+	"strings"
+)
+
+// FilterKind is used to specify whether a filter should be applied to
+// logs that match the filter or logs that do not match the filter.
+type FilterKind int
+
+// FilterKind values
+const (
+	FilterIn FilterKind = iota
+	FilterOut
+)
+
+// LoggerFilter is used to filter logs.
+type LoggerFilter struct {
+	msg              *filterStringContains
+	pkg              *filterString
+	file             *filterString
+	lvl              *filterInt
+	msgRegex         *filterStringRegex
+	filterInEnabled  bool
+	filterOutEnabled bool
+}
+
+// NewLoggerFilter creates a new LoggerFilter.
+func NewLoggerFilter() LoggerFilter {
+	return LoggerFilter{
+		msg:              newFilterStringContains(),
+		pkg:              newFilterString(),
+		file:             newFilterString(),
+		lvl:              newFilterInt(),
+		msgRegex:         newFilterStringRegex(),
+		filterInEnabled:  false,
+		filterOutEnabled: false,
+	}
+}
+
+// AddMsg adds a filter value and type to be checked against messages.
+func (lf *LoggerFilter) AddMsg(s string, t FilterKind) {
+	lf.msg.add(s, t)
+	lf.setFilterEnabled(t)
+}
+
+// AddPkg adds a filter value and type to be checked against package names.
+func (lf *LoggerFilter) AddPkg(s string, t FilterKind) {
+	lf.pkg.add(s, t)
+	lf.setFilterEnabled(t)
+}
+
+// AddFile adds a filter value and type to be checked against file names.
+func (lf *LoggerFilter) AddFile(s string, t FilterKind) {
+	lf.file.add(s, t)
+	lf.setFilterEnabled(t)
+}
+
+// AddLvl adds a filter value and type to be checked against log levels.
+func (lf *LoggerFilter) AddLvl(i int, t FilterKind) {
+	lf.lvl.add(i, t)
+	lf.setFilterEnabled(t)
+}
+
+// AddMsgRegex adds a filter regex and type to be checked against messages.
+func (lf *LoggerFilter) AddMsgRegex(regex string, t FilterKind) error {
+	if err := lf.msgRegex.add(regex, t); err != nil {
+		return err
+	}
+	lf.setFilterEnabled(t)
+
+	return nil
+}
+
+// Enabled returns true if any filters are enabled.
+func (lf *LoggerFilter) Enabled() bool {
+	return lf.filterInEnabled || lf.filterOutEnabled
+}
+
+// setFilterEnabled sets the filter enabled flag for the given filter type.
+func (lf *LoggerFilter) setFilterEnabled(t FilterKind) {
+	if t == FilterIn {
+		lf.filterInEnabled = true
+	} else {
+		lf.filterOutEnabled = true
+	}
+}
+
+// filterString
+
+type filterString struct {
+	vals map[string]FilterKind
+}
+
+func newFilterString() *filterString {
+	return &filterString{
+		vals: map[string]FilterKind{},
+	}
+}
+
+func (fs *filterString) add(s string, t FilterKind) {
+	if _, ok := fs.vals[s]; ok && t == FilterIn {
```

Comment:
if I understood correctly the idea here is that we avoid overriding an existing filter with a filter-out? if so, maybe log a debug message whenever an override happens? could help a user identify a mistake

Reply:
> if I understood correctly the idea here is that we avoid overriding an existing filter with a filter-out? 

Exactly. Since filter-out has precedence.

> If so, maybe log a debug message whenever an override would happen? could help a user identify a mistake.

Good idea. 👍🏼

Reply:
Oh, wait... we are still at first stage building the logger. Trying to log something would work correctly only if via stderr. `fmt.Fprintf(os.Stderr...)` WDYT?

Reply:
oh! true. Maybe leave as is than

Reply:
I'll think about it more.

Reply:
Ok, I created the `ErrFilterOutExistsForKey` error to be checked and logged instead of being considered a failure.